catalog/app/components/Dialog/Prompt.tsx

```diff
@@ -0,0 +1,115 @@
+import * as R from 'ramda'
+import * as React from 'react'
+import * as M from '@material-ui/core'
+import * as Lab from '@material-ui/lab'
+
+interface DialogProps {
+  initialValue?: string
+  onCancel: () => void
+  onSubmit: (value: string) => void
+  open: boolean
+  title: string
+  validate: (value: string) => Error | undefined
+}
+
+function Dialog({
+  initialValue,
+  open,
+  onCancel,
+  onSubmit,
+  title,
+  validate,
+}: DialogProps) {
+  const [value, setValue] = React.useState(initialValue || '')
+  const [submited, setSubmited] = React.useState(false)
+  const error = React.useMemo(() => validate(value), [validate, value])
+  const handleChange = React.useCallback((event) => setValue(event.target.value), [])
+  const handleSubmit = React.useCallback(
+    (event) => {
+      event.preventDefault()
+      setSubmited(true)
+      if (!error) onSubmit(value)
+    },
+    [error, onSubmit, value],
+  )
+  return (
+    <M.Dialog open={open} fullWidth maxWidth="sm">
+      <form onSubmit={handleSubmit}>
+        <M.DialogTitle>{title}</M.DialogTitle>
+        <M.DialogContent>
+          <M.TextField
+            autoFocus
+            fullWidth
+            margin="dense"
+            onChange={handleChange}
+            value={value}
+          />
+          {!!error && !!submited && (
+            <Lab.Alert severity="error">{error.message}</Lab.Alert>
+          )}
+        </M.DialogContent>
+        <M.DialogActions>
+          <M.Button onClick={onCancel} color="primary" variant="outlined">
+            Cancel
+          </M.Button>
+          <M.Button
+            color="primary"
+            disabled={!!error && !!submited}
+            onClick={handleSubmit}
+            variant="contained"
+          >
+            Submit
+          </M.Button>
+        </M.DialogActions>
+      </form>
+    </M.Dialog>
+  )
+}
+
+interface PromptProps {
+  initialValue?: string
+  onSubmit: (value: string) => void
+  title: string
+  validate: (value: string) => Error | undefined
+}
+
+export function usePrompt({ initialValue, title, onSubmit, validate }: PromptProps) {
+  const [key, setKey] = React.useState(0)
+  const [opened, setOpened] = React.useState(false)
+  const open = React.useCallback(() => setOpened(true), [])
+  const close = React.useCallback(() => {
+    setOpened(false)
+    setKey(R.inc)
```

Comment:
this would probably cause flash of the dialog contents while it's in closing transition.
solution to this would be doing key change either after the transition is complete (which would complicate the code with additional state and callbacks) or before the dialog is shown (before `setOpened(true)`)

Reply:
> before the dialog is shown

:thinking: